src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ClassWriter.java

```diff
@@ -724,5 +732,27 @@ protected void writeClassAttributes() throws IOException {
                 dos.writeShort(innerClasses.at(index));
             }
         }
+
+        // write bootstrap method attribute, if any
+        if (bsmCount != 0) {
+            ConstantPool cpool = klass.getConstants();
+            writeIndex(_bootstrapMethodsIndex);
+            if (DEBUG) debugMessage("bootstrap methods attribute = " + _bootstrapMethodsIndex);
+            int attrLen = 2; // num_bootstrap_methods
+            for (int index = 0; index < bsmCount; index++) {
+                int bsmArgsCount = cpool.getBootstrapMethodArgsCount(index);
+                attrLen += 2 // bootstrap_method_ref
+                           + 2 // num_bootstrap_arguments
+                           + bsmArgsCount * 2;
+            }
+            dos.writeInt(attrLen);
+            dos.writeShort(bsmCount);
+            for (int index = 0; index < bsmCount; index++) {
+                short value[] = cpool.getBootstrapMethodAt(index);
+                for (int i = 0; i < value.length; i++) {
+                    dos.writeShort(value[i]);
+                }
+            }
+        }
```

Comment:
Hi, it looks odd to me that we only write one short in the loop: should that be a bootstrap method ref, num arguments, and arguments?

I was comparing with: https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf
4.7.23 The BootstrapMethods Attribute

We have bootstrap attribute written by other code, e.g.
"src/jdk.jdeps/share/classes/com/sun/tools/classfile/ClassWriter.java" 
visitBootstrapMethods

"src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassWriter.java" 
writeBootstrapMethods

Reply:
This is because the array returned by getBootstrapMethodAt() includes bootstrap method ref, num arguments and arguments. getBootstrapMethodAt() gets all the pieces and assembles them in the array which is then written out in this loop. There is already getBootstrapSpecifierAt() which is doing the same thing. So I just reused that implementation.

Reply:
ah yes thanks, hadn't got there yet 8-)